[material-ui][Rating] Deprecate *Props and complete `slots`, `slotProps`

harry-whorlow opened the pull request:
part of https://github.com/mui/material-ui/issues/41281

This PR adds slot the slot prop to rating, whilst deprecating components.

props.iconContainerComponent -> slot.iconContainer


[/] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

harry-whorlow commented on packages/mui-material/src/Rating/Rating.d.ts:
```diff
@@ -4,14 +4,54 @@ import { OverridableStringUnion } from '@mui/types';
+>;
 
-export interface RatingOwnProps {
+export interface RatingOwnProps extends RatingSlotsAndSlotProps {
```
@DiegoAndai, no still the same 🙃

DiegoAndai replied:
It has to be added to the `RatingTypeMap`, not `RatingOwnProps`, does it work with that change?

harry-whorlow replied:
yeah, thats what I did here is the commit changes [dab6cfb](https://github.com/mui/material-ui/pull/45295/commits/dab6cfba91e039edff3b021df2a6491df7c3daa0)

siriwatknp replied:
The proptypes generation depends on the implementation, not the .d.ts files. It's now updated.

siriwatknp commented:
@harry-whorlow Thanks for your contribution. I was working on this before I went on holiday last week, so let me take over this PR.

Note, the `IconContainerComponent` is not a slot (based on the prior code), it should be updated to `slotProps.icon.component` instead.

harry-whorlow commented:
@siriwatknp ah I see, sure thing take it away... 

Out of curiosity, was the reason the pnpm proptypes not updating was because I needed to extend `RatingOwnProps` *and* the `RatingTypeMap` instead of one or the other?

DiegoAndai commented on packages/mui-material/src/Rating/Rating.js:
```diff
@@ -524,25 +550,54 @@ const Rating = React.forwardRef(function Rating(inProps, ref) {
-      onMouseLeave={handleMouseLeave}
-      className={clsx(
-        classes.root,
-        {
-          // TODO v6: remove this class as it duplicates with the global state class Mui-readOnly
-          'MuiRating-readOnly': readOnly,
-        },
```
I agree with removing this, but we should add it to the upgrade guide: https://github.com/mui/material-ui/blob/master/docs/data/material/migration/upgrade-to-v7/upgrade-to-v7.md

DiegoAndai reviewed (changes requested):
Hey @harry-whorlow, looks good. We need to add this to the deprecated APIs guide: https://github.com/mui/material-ui/blob/master/docs/data/material/migration/migrating-from-deprecated-apis/migrating-from-deprecated-apis.md

harry-whorlow commented:
@DiegoAndai, thank's but @siriwatknp deserves the credit on this one! I've rebased master and updated the migration to note the removal of `MuiRating-readOnly`.  🤟
